src/snowflake/snowpark/dataframe.py

```diff
@@ -2060,6 +2060,21 @@ def join(
                 using_columns = [using_columns]
             elif isinstance(using_columns, Column):
                 using_columns = using_columns
+            elif (
+                isinstance(using_columns, list)
+                and len(using_columns) > 0
+                and not all([type(col) == str for col in using_columns])
+            ):
+                bad_col, bad_idx = None, 0
+                for idx, col in enumerate(using_columns):
+                    if type(col) != str:
+                        bad_col = col
+                        bad_idx = idx
+                        break
+                raise TypeError(
+                    f"All list elements for 'on' or 'using_columns' must be string type. "
```

Comment:
Does it accept a `Column` object?

Reply:
If `using_column` is not `str` or `Column`, the subsequent code is written in such a way that works well only if the elements of iterable are of type `str`.